File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return head
            
        tail = head
        length = 1

        #Step 1: Reach tail & also find len of LL at the same time
        while tail.next:
            length += 1
            tail = tail.next
        
        #Connecr tail to head
        tail.next =head
        
        #Modify k to handle cases where k >= len(LL)
        k %= length

        #Step 2: Reach node = len-k
        temp = head
        for _ in range(1,length-k):
            temp = temp.next
        
        #Step 3: Update head
        head = temp.next

        #Step 4: Break the link betn head & temp
        temp.next = None

        return head
        

File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null){
            return head;
        }

        ListNode curr = head;
        ListNode prev = null;

        k = k % getLength(head);

        for(int i = 0; i < k; i++){
            while(curr.next != null){
                prev = curr;
                curr = curr.next;
            }
            curr.next = head;
            head = curr;
            prev.next = null;
        }
        return head;
    }

    private int getLength(ListNode head){
        ListNode curr = head;
        int len = 0;
        while(curr != null){
            len++;
            curr = curr.next;
        }
        return len;
    }
}